src/libsyntax/ext/expand.rs

```diff
@@ -121,6 +121,47 @@ impl MacResult for MacroScopePlaceholder {
     }
 }
 
+// If keep_macs is true, this is the result of a MacroRulesTT expansion:
+// a struct that can recreate the macro definition.
+struct MacroDefExpn {
```

Comment:
Could you use `MacEager::items` instead of defining this struct?
`MacEager::items(SmallVector::one(P(ast::Item { ... })))` is a `Box<MacResult>` for which `make_items` returns the given item.